lib/scan/whatcms/engine.py

```diff
@@ -0,0 +1,297 @@
+from core.load_modules import load_file_path
+from lib.socks_resolver.engine import getaddrinfo
+from core._time import now
+from core.log import __log_into_file
+import requests
+
+
+def extra_requirements_dict():
+    return {
+        "whatcms_ports": [443],
+        "whatcms_api_key": ["test_api_key"],
+    }
+
+
+def conn(targ, port, timeout_sec, socks_proxy):
+    try:
+        if socks_proxy is not None:
+            socks_version = (
+                socks.SOCKS5
+                if socks_proxy.startswith("socks5://")
+                else socks.SOCKS4
+            )
+            socks_proxy = socks_proxy.rsplit("://")[1]
+            if "@" in socks_proxy:
+                socks_username = socks_proxy.rsplit(":")[0]
+                socks_password = socks_proxy.rsplit(":")[1].rsplit("@")[0]
+                socks.set_default_proxy(
+                    socks_version,
+                    str(socks_proxy.rsplit("@")[1].rsplit(":")[0]),
+                    int(socks_proxy.rsplit(":")[-1]),
+                    username=socks_username,
+                    password=socks_password,
+                )
+                socket.socket = socks.socksocket
+                socket.getaddrinfo = getaddrinfo
+            else:
+                socks.set_default_proxy(
+                    socks_version,
+                    str(socks_proxy.rsplit(":")[0]),
+                    int(socks_proxy.rsplit(":")[1]),
+                )
+                socket.socket = socks.socksocket
+                socket.getaddrinfo = getaddrinfo()
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sys.stdout.flush()
+        s.settimeout(timeout_sec)
+        s.connect((targ, port))
+        return s
+    except Exception:
+        return None
+
+
+def whatcms(
+    target,
+    port,
+    timeout_sec,
+    log_in_file,
+    language,
+    time_sleep,
+    thread_tmp_filename,
+    socks_proxy,
+    scan_id,
+    scan_cmd,
+    whatcms_api_key,
+):
+    try:
+        try:
+            s = conn(target, port, timeout_sec, socks_proxy)
+        except Exception:
+            return False
+        if not s:
+            return False
+        else:
+            global cms_name
+            headers = {
+                "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:77.0) Gecko/20100101 Firefox/77.0",
+                "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
+                "Accept-Language": "en-US,en;q=0.7,ru;q=0.3",
+                "Accept-Encoding": "gzip, deflate",
+                "Connection": "Keep-Alive",
+                "Upgrade-Insecure-Requests": "1",
+                "X-Requested-With": "XMLHttpRequest",
+            }
+            requests_url = (
+                "https://whatcms.org/API/CMS?key="
+                + whatcms_api_key
+                + "&url="
+                + target
+            )
+            req = requests.get(requests_url, verify=False, headers=headers)
```

Comment:
Error Handling is missing - should check for response HTTP code and show display an error messge